### ISSUE TITLE
Composer install resolves messed

### ISSUE THREAD

**Issue by cyrilchapon** [composer/composer#5721]
Hello.

I've just landed in a SYmfony project and I'm struggling composer.
I've got a package (**friendsofsymfony/rest-bundle**) which version is set to "1.8.0", but running `composer install` resolves it to *1.8.0* on my local machine (windows), but to *2.1.0* on my Debian server.

Local `composer --version`

```
Composer version 1.2.1 2016-09-12 11:27:19
```

Server `composer --version`

```
Composer version 1.2.0 2016-07-19 01:28:52
```

My `composer.json`:
<details>
<summary>CLICK TO EXPAND</summary>
```json
{
    "name": "hidden",
    "license": "MIT",
    "type": "project",
    "description": "hidden",
    "autoload": {
        "psr-4": { "": "src/" },
        "classmap": [ "app/AppKernel.php", "app/AppCache.php" ]
    },
    "require": {
        "php": ">=5.3.9",
        "symfony/symfony": "2.8.*",
        "doctrine/orm": "^2.4.8",
        "doctrine/doctrine-bundle": "~1.4",
        "symfony/swiftmailer-bundle": "~2.3",
        "symfony/monolog-bundle": "~2.4",
        "sensio/distribution-bundle": "~5.0",
        "sensio/framework-extra-bundle": "^3.0.2",
        "incenteev/composer-parameter-handler": "~2.0",
        "friendsofsymfony/rest-bundle": "1.8.0",
        "jms/serializer-bundle": "1.1.0",
        "doctrine/doctrine-fixtures-bundle": "~2.3",
        "symfony/http-foundation": "~3.1",
        "symfony/var-dumper": "^2.8",
        "phpoffice/phpexcel": "^1.8",
        "symfony/filesystem": "^2.8",
        "twig/extensions": "^1.3",
        "friendsofsymfony/message": "1.0.x-dev",
        "knplabs/knp-time-bundle": "^1.6",
        "jasongrimes/paginator": "^1.0",
        "cmfcmf/openweathermap-php-api": "^2.1",
        "friendsofsymfony/jsrouting-bundle": "^1.6"
    },
    "require-dev": {
        "sensio/generator-bundle": "~3.0",
        "symfony/phpunit-bridge": "~2.7",
        "phpunit/phpunit": "5.3.*"
    },
    "scripts": {
        "post-install-cmd": [
            "Incenteev\\ParameterHandler\\ScriptHandler::buildParameters",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::buildBootstrap",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::clearCache",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installAssets",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installRequirementsFile",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::prepareDeploymentTarget"
        ],
        "post-update-cmd": [
            "Incenteev\\ParameterHandler\\ScriptHandler::buildParameters",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::buildBootstrap",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::clearCache",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installAssets",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installRequirementsFile",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::prepareDeploymentTarget"
        ]
    },
    "config": {
        "bin-dir": "bin"
    },
    "extra": {
        "symfony-app-dir": "app",
        "symfony-web-dir": "web",
        "symfony-assets-install": "relative",
        "incenteev-parameters": {
            "file": "app/config/parameters.yml"
        }
    }
}

```
</details>

Output of `composer diagnose`:
<details>
<summary>CLICK TO EXPAND</summary>
```
Checking composer.json: WARNING
Defining autoload.psr-4 with an empty namespace prefix is a bad idea for performance
require.friendsofsymfony/rest-bundle : exact version constraints (1.8.0) should be avoided if the package follows semantic versioning
require.jms/serializer-bundle : exact version constraints (1.1.0) should be avoided if the package follows semantic versioning
require.friendsofsymfony/message : exact version constraints (1.0.x-dev) should be avoided if the package follows semantic versioning
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com rate limit: OK
Checking disk free space: OK
Checking pubkeys:
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
```
</details>

Some facts may be worth noting :
- My project has stopped working on my local machine when I did this :
  - (I had 1.8.0 inside my composer.json, but I don't know which version was actually resolved. I tried to update to 2.1.0)
  - `composer require friendsofsymfony/rest-bundle:2.1.0`
  - (I noticed this had too many side effects, so I decided to revert)
  - `composer require friendsofsymfony/rest-bundle:1.8.0`
  - Noticed my project doesn't work anymore (the dependency throws an error it didn't throw)
  - Tried removing ./vender, clearing composer cache, uninstalling/reinstalling composer, rechecking-out my entire project then `composer install`, but my project still doesn't work
  - Now if I run `composer require friendsofsymfony/rest-bundle:2.1.0`, just for fun, this doesn't work anymore
```
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - friendsofsymfony/rest-bundle 2.1.0 conflicts with sensio/framework-extra-bundle[v3.0.12].
    - friendsofsymfony/rest-bundle 2.1.0 conflicts with sensio/framework-extra-bundle[v3.0.12].
    - friendsofsymfony/rest-bundle 2.1.0 conflicts with sensio/framework-extra-bundle[v3.0.12].
    - Installation request for friendsofsymfony/rest-bundle 2.1.0 -> satisfiable by friendsofsymfony/rest-bundle[2.1.0].
    - Installation request for sensio/framework-extra-bundle (locked at v3.0.12, required as ^3.0.2) -> satisfiable by sensio/framework-extra-bundle[v3.0.12].


Installation failed, reverting ./composer.json to its original content.
```
  - Notice that I never commited any composer.json change. So the fact is when I `composer install`ed my project with the provided composer.json, a week ago, I had a release of this dependency, but now I've got another.
  - If I copy whole *vendor* folder (including rest-bundle resolved to 2.1.0) from my server to my local, my project works again (which tends to tell me that the initial local resolved version was 2.1.0)

**Comment by stof** [comment]
well, if you want to install ``friendsofsymfony/rest-bundle:2.1.0``, you will have to allow composer to change the version of ``sensio/framework-extra-bundle`` being used, as FOSRestBundle 2.1.0 conflicts with the version you have (and ``composer require`` locks other packages).
You should change the requirement in your composer.json and perform a partial update involving multiple packages.

Btw, if you have ``1.8.0`` in your composer.json, the resolved version will be 1.8.0, as it is the only version which can match such requirement (as it is an exact match constraint)

**Comment by cyrilchapon** [comment]
Hey @stof. Thanks for your answer

> Btw, if you have 1.8.0 in your composer.json, the resolved version will be 1.8.0, as it is the only version which can match such requirement (as it is an exact match constraint)

Yes it should. But this is not the case:

*Composer.json* on my server

```
[...]
"friendsofsymfony/rest-bundle": "1.8.0"
[...]
```

*vendor/friendsofsymfony/rest-bundle* on my server :

```
    "extra": {
        "branch-alias": {
            "dev-master": "2.1-dev"
        }
    }
```

**Comment by cyrilchapon** [comment]
My mistake. Team misunderstanding. Composer.json was commited.
Thanks for your help.
